x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/action/profile/TransportProfileHasPrivilegesActionTests.java

```diff
@@ -180,10 +190,84 @@ public void testNoProfileSubjectsFound() throws Exception {
+                .indices(randomAlphaOfLengthBetween(2, 8))
+                .query(new BytesArray(randomAlphaOfLength(5)))
+                .build();
+        }
+
+        final ProfileHasPrivilegesRequest request = new ProfileHasPrivilegesRequest(
+            randomList(1, 100, () -> randomAlphaOfLengthBetween(4, 10)),
+            new AuthorizationEngine.PrivilegesToCheck(
+                new String[0],
+                indicesPrivileges,
+                new RoleDescriptor.ApplicationResourcePrivileges[0],
+                randomBoolean()
+            )
+        );
+
+        final PlainActionFuture<ProfileHasPrivilegesResponse> listener = new PlainActionFuture<>();
+        transportProfileHasPrivilegesAction.execute(mock(CancellableTask.class), request, listener);
+
+        final IllegalArgumentException ile = expectThrows(IllegalArgumentException.class, () -> listener.actionGet());
+        assertThat(ile, notNullValue());
+        assertThat(ile.getMessage(), containsString("may only check index privileges without any DLS query"));
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testCancellation() {
+        final List<String> profileUids = new ArrayList<>(new HashSet<>(randomList(1, 5, () -> randomAlphaOfLengthBetween(5, 10))));
+        final ProfileHasPrivilegesRequest request = new ProfileHasPrivilegesRequest(profileUids, randomValidPrivilegesToCheckRequest());
+        doAnswer(invocation -> {
+            Collection<String> uidsArg = (Collection<String>) invocation.getArguments()[0];
+            Map<String, Subject> profileUidToSubject = new HashMap<>();
+            for (String uid : uidsArg) {
+                profileUidToSubject.put(uid, new Subject(new User("user_for_profile_" + uid), mock(Authentication.RealmRef.class)));
+            }
+            final ActionListener<MultiProfileSubjectResponse> listener = (ActionListener<MultiProfileSubjectResponse>) invocation
+                .getArguments()[1];
+            listener.onResponse(new MultiProfileSubjectResponse(profileUidToSubject, Set.of()));
+            return null;
+        }).when(profileService).getProfileSubjects(anyCollection(), anyActionListener());
+
+        doAnswer(invocation -> {
+            final ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener = (ActionListener<
+                Collection<ApplicationPrivilegeDescriptor>>) invocation.getArguments()[2];
+            listener.onResponse(List.of());
+            return null;
+        }).when(nativePrivilegeStore).getPrivileges(anyCollection(), any(), anyActionListener());
+
+        final AtomicInteger cancelAfter = new AtomicInteger(randomIntBetween(1, profileUids.size()));
+        final CancellableTask cancellableTask = new CancellableTask(0, "type", "action", "description", TaskId.EMPTY_TASK_ID, Map.of());
+
+        if (cancelAfter.decrementAndGet() == 0) {
+            TaskCancelHelper.cancel(cancellableTask, "reason");
+        }
```

Comment:
I am a little unsure about the intention of this new `if` block. Is it to address my previous comment about how `isCancelled` is `not` checked in the main code when `partialResultSize == profileUids.size()`. I thought we agreed it is a valid randomisation scenario and we should keep it? 

Or are you trying to cover the scenario where the task is cancelled before even the 1st check (I believe that's what happens when `profileUids.size() == 1`? This is a valid scenario as well. But it makes the previous scenario (cancelled _after_ last check is performed) no longer tested. Would it be better to wrap this with a `randomBoolean`, i.e. `if (randomBoolean() && cancelAfter.decrementAndGet() == 0) { ... }`, to maximise its coverage for both scenarios?

Reply:
I have pushed [a39a476](https://github.com/elastic/elasticsearch/pull/87224/commits/a39a4768d975869cfed7626b11d8a35e8a4a7b57) .

The task in the test scenario is now cancelled before any `checkPrivileges` as well as after all the `checkPrivileges` .